pkg/vault/test/foundry/E2eSwap.t.sol

```diff
@@ -601,14 +674,20 @@ contract E2eSwapTest is BaseVaultTest {
         poolConfig.tokenDecimalDiffs = PoolConfigLib.toTokenDecimalDiffs(tokenDecimalDiffs);
         vault.manualSetPoolConfig(pool, poolConfig);
 
-        // Fix pool init amounts, adjusting to new decimals. These values will be used to calculate max swap values and
-        // pool liquidity.
-        poolInitAmountTokenA = poolInitAmount.mulDown(10 ** decimalsTokenA);
-        poolInitAmountTokenB = poolInitAmount.mulDown(10 ** decimalsTokenB);
-
-        _setPoolBalances(poolInitAmountTokenA, poolInitAmountTokenB);
+        setPoolInitAmounts();
+        setPoolBalances(poolInitAmountTokenA, poolInitAmountTokenB);
 
         // Min and Max swap amounts depends on the decimals of each token, so a recalculation is needed.
         calculateMinAndMaxSwapAmounts();
     }
+
+    function setPoolInitAmounts() internal {
+        uint256 rateTokenA = getRate(tokenA);
+        uint256 rateTokenB = getRate(tokenB);
+
+        // Fix pool init amounts, adjusting to new decimals. These values will be used to calculate max swap values and
+        // pool liquidity.
+        poolInitAmountTokenA = poolInitAmount.mulDown(10 ** (decimalsTokenA)).divDown(rateTokenA);
+        poolInitAmountTokenB = poolInitAmount.mulDown(10 ** (decimalsTokenB)).divDown(rateTokenB);
```

Comment:
Is this to start with equal value? Do we really need this for the test to work?

Reply:
PoolMock doesn't need this since the math is linear (it doesn't matter what the current pool liquidity is; the result will be the same). This is required, especially for PoolWeighted, and sometimes for PoolStable, because the math behaves differently near the edges. So, putting the pool near the middle of the price curve ensures a more predictable result of the test, generically. (I mean, we could predict using the pool math, but then the test wouldn't be as generic)

Reply:
I think we should test the edges.

We don't really need to predict the exact result. Relative error can be a bit more loose to the 'lossy' side: if a user loses a bit more because of rounding when doing / undoing something it's still fine. What we don't want is the user getting any value out of the operations for free, and we check that with `Gt` / `Lt` comparing user balances and invariants.